### PR TITLE
A CLI tool to assist during an upgrade to OpenSearch.

### LINKED ISSUE #780: [UPGRADE] Path for properties fetched via system.getProperty
**Is your feature request related to a problem? Please describe.**

We have renamed properties, e.g. `es.distribution.type` has become `opensearch.distribution.type`. What is the path forward for those settings?

**Describe the solution you'd like**

Document the way this breaks for users. 
Provide a migration path.
Optionally provide tooling to aid users in the migration path.

**Additional context**
Coming from #640.

### PR DESCRIPTION
**Pull request by adnapibar**
### Description

This change adds an **initial version** of a new CLI tool `opensearch-upgrade`  into the OpenSearch distribution. This tool will help during an upgrade of a node running Elasticsearch `v7.10.2/v6.8.0` to OpenSearch 1.x. It automates the process of importing existing configurations and installing official plugins (*currently, it won't work as the plugin artifacts are yet to be published*).

The tool can be run using a single command, `$ES_HOME/bin/opensearch-upgrade` , which performs the tasks of configuring the OpenSearch installation as part of the upgrading the node to OpenSearch. 

### Issues Resolved
Resolves #780 #781 #782 #785 #787 
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Signed-off-by: Rabi Panda <adnapibar@gmail.com>

### CODE REVIEW

File: distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/ImportLog4jPropertiesTask.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+
+/**
+ * Imports Log4j properties from an existing elasticsearch installation.
+ */
+class ImportLog4jPropertiesTask implements UpgradeTask {
```
Comment (adnapibar) [inline distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/ImportLog4jPropertiesTask.java:22]: This needs to be updated to only import any additional settings and not copy and overwrite the entire file.

**Comment by saratvemulapalli** [comment]
Love this, I haven't completely read through the PR. 
But a question, if a repository outside of OpenSearch eg: Plugins, clients etc have their static configuration files. 
Can we extend this for them?

**Comment by saratvemulapalli** [comment]
start gradle check

**Comment by adnapibar** [comment]
> Love this, I haven't completely read through the PR.
> But a question, if a repository outside of OpenSearch eg: Plugins, clients etc have their static configuration files.
> Can we extend this for them?

Not sure I understand the question completely - can you give an example of a such a scenario?

**Comment by saratvemulapalli** [comment]
> > Love this, I haven't completely read through the PR.
> > But a question, if a repository outside of OpenSearch eg: Plugins, clients etc have their static configuration files.
> > Can we extend this for them?
> 
> Not sure I understand the question completely - can you give an example of a such a scenario?

Sure definitely. Let's take Opendistro security plugin as an example. It has bunch of config files and also has config residing in elasticsearch.yml. These parameters have been updated in the opensearch plugin and customers would need to manually translate this configuration. 
So the question was, can this tool be extended for use cases like these?
Ref: https://opendistro.github.io/for-elasticsearch-docs/docs/security/configuration/yaml/

**Comment by adnapibar** [comment]
> Sure definitely. Let's take Opendistro security plugin as an example. It has bunch of config files and also has config residing in elasticsearch.yml. These parameters have been updated in the opensearch plugin and customers would need to manually translate this configuration.
> So the question was, can this tool be extended for use cases like these?
> Ref: https://opendistro.github.io/for-elasticsearch-docs/docs/security/configuration/yaml/

Thanks. Yes, we can certainly extend the tool for importing these.

**Comment by nknize** [comment]
start gradle check

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
I think this is great, we can improve from here.

**Comment by adnapibar** [comment]
start gradle check

**Pull request merged**